apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerNumeric.cs

```diff
@@ -379,27 +414,76 @@ public static SpannerNumeric FromDecimal(decimal value, LossOfPrecisionHandling
 
         /// <summary>
-        /// Converts this value to <see cref="SpannerNumeric"/>, 
+        /// Converts this value to <see cref="decimal"/>, 
         /// </summary>
         /// <remarks>
         /// This conversion may silently lose precision, depending on <paramref name="lossOfPrecisionHandling"/>, but 
-        /// will always throw <see cref="OverflowException"/> if value is out of the range of <see cref="Decimal"/>.
+        /// will always throw <see cref="OverflowException"/> if value is out of the range of <see cref="decimal"/>.
         /// </remarks>
         /// <param name="lossOfPrecisionHandling">How to handle values with signficant digits that would be lost by the conversion.</param>
-        /// <exception cref="OverflowException">This value is outside the range of <see cref="Decimal"/>.</exception>
+        /// <exception cref="OverflowException">This value is outside the range of <see cref="decimal"/>.</exception>
         /// <returns>The converted value.</returns>
         public decimal ToDecimal(LossOfPrecisionHandling lossOfPrecisionHandling)
         {
-            // FIXME: Awful performance!
-            string text = ToString();
+            if (_integer == BigInteger.Zero)
+            {
+                return 0m;
+            }
+
             // Handles overflow
-            decimal dec = Decimal.Parse(text, CultureInfo.InvariantCulture);
-            if (lossOfPrecisionHandling == LossOfPrecisionHandling.Throw && this != (SpannerNumeric) dec)
+            return decimal.Parse(ToString(DecimalPrecision), CultureInfo.InvariantCulture);
+
+            int DecimalPrecision(string scaledText, int integerPlaces, int decimalPlaces)
             {
-                // TODO: Is this the right exception to use?
-                throw new InvalidOperationException("Conversion would lose information");
+                int decimalPrecision;
+                // If the number had no decimal places to begin with, then we need not to worry.
+                // If the number has 29 or more integer places we cannot include any decimal places;
+                // this may overflow when we convert to decimal but that's fine, let's not handle that here.
+                if (decimalPlaces == 0 || integerPlaces >= 29)
+                {
+                    decimalPrecision = 0;
+                }
+                // If the number has at most 28 significant digits we can represent it fully.
+                // Note that this takes of the case where there are no integer places as well,
+                // since decimal places for SpannerNumeric are at most 9.
+                else if (integerPlaces + decimalPlaces <= 28)
+                {
+                    decimalPrecision = decimalPlaces;
+                }
+                // integerPlaces + decimalPlaces >= 29 and
+                // 1 <= integerPlaces <= 28 which means we may have room for some of the decimal places
```

Comment:
integerPlaces could be zero here - we don't have the same check that's in PgNumeric. That may be a correct code difference, but I think it means the comment is incorrect.

Reply:
Nope, integerPlaces being zero would have been solved on the `else if` above, as the maximum number of decimal places on `SpannerNumeric` is 9, so:
```
integerPlaces == 0 => integerPlaces + decimalPlaces <= 9 <= 28
```

There's a note on the `else if` above but let me know if that's not clear enough.

Reply:
Ah yes - I hadn't read it carefully enough.